packages/terra-application-utility/src/header/_HeaderUtilityMenu.jsx

```diff
@@ -0,0 +1,256 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import classNames from 'classnames/bind';
+import 'terra-base/lib/baseStyles';
+import ContentContainer from 'terra-content-container';
+import IconLeft from 'terra-icon/lib/icon/IconLeft';
+import MenuDivider from '../_UtilityMenuDivider';
+import Utils from '../_Utils';
+import MenuItem from './_HeaderUtilityMenuItem';
+import styles from './_HeaderUtilityMenu.scss';
+
+const cx = classNames.bind(styles);
+
+const propTypes = {
+  /**
+   * Key of the top level menu.
+   */
+  initialSelectedKey: PropTypes.string.isRequired,
+  /**
+   * Indicates if the height is bound to a value.
+   */
+  isHeightBounded: PropTypes.bool,
+  /**
+   * The config file containing the menu items to be rendered.
+   */
+  menuItems: PropTypes.arrayOf((Utils.itemShape)).isRequired,
+  /**
+   * The function to trigger when a menu item is selected.
+   */
+  onChange: PropTypes.func.isRequired,
+  /**
+   * The function that closes the menu.
+   */
+  onRequestClose: PropTypes.func,
+};
+
+const defaultProps = {
+  isHeightBounded: undefined,
+};
+
+const processMenuItems = (items) => {
+  const map = new Map();
+  items.forEach((item) => {
+    map.set(
+      item.key,
+      { id: item.id,
+        itemKey: item.key,
+        title: item.title,
+        content: item.content,
+        contentLocation: item.contentLocation,
+        isSelected: item.isSelected,
+        isSelectable: item.isSelectable,
+        childKeys: item.childKeys,
+      },
+    );
+  });
+  return map;
+};
+
+const hasChevron = (item) => {
+  if (item.childKeys && item.childKeys.length > 0) {
+    return true;
+  }
+  return false;
+};
+
+class HeaderUtilityMenu extends React.Component {
+  constructor(props) {
+    super(props);
+    this.getItem = this.getItem.bind(this);
+    this.handleOnChange = this.handleOnChange.bind(this);
+    this.handleRequestBack = this.handleRequestBack.bind(this);
+    this.pop = this.pop.bind(this);
+    this.push = this.push.bind(this);
+    this.toggleIsSelected = this.toggleIsSelected.bind(this);
+    this.state = {
+      map: processMenuItems(props.menuItems),
+      currentKey: props.initialSelectedKey,
+      previousKeyStack: [],
+    };
+  }
+
+  componentWillReceiveProps(nextProps) {
+    this.setState({
+      map: processMenuItems(nextProps.menuItems),
```

Comment:
You could do a simple equality check around this to prevent unnecessary state updates if your consumers are using immutable props:

```
if (nextProps.menuItems === this.props.menuItems) { 
  return;
}
```